### PR TITLE
[WIP] Expose BasicDeliverEventArgs in message context

### LINKED ISSUE #103: Propagate all BasicDeliverEventArgs arguments
When sending a message from native RabbitMQ API to a NSB endpoint not all BasicDeliverEventArgs are being propogated into the bus.CurrentMessageContext object.

These object could be Exchange, RoutingKey, UserID, etc....

Why do I need these parameters: In an IoT environment where non .NET devices connects directly to the broker there is a need to know which device sends the message. The device code can add a property to the message identifying who it is but this is not secure because a compromised code can impersonate to another device. We need to find a way that the broker infrastructure will identify the client.

One of the options to do it is to use client side certificate or any other rabbit routing mechanism. All these solutions eventually mapped to one or more parameters in the BasicDeliverEventArgs or in BasicDeliverEventArgs.BasicProperties. Unfortunately these parameters are not copied into the TransportMessage object (and eventually into the bus.CurrentMessageContext object) at MessageConverter.ToTransportMessage method

### PR DESCRIPTION
**Pull request by adamralph**
fixes #103

### CODE REVIEW

**Comment by adamralph** [comment]
@Particular/rabbitmq-transport-maintainers this is now ready to go, but as @bording and I discussed, we need to think about whether this is really how we want things to work. Another option available to us is to pick out the interesting data that is available on `BasicDeliverEventArgs` and put them into appropriate headers.

**Comment by adamralph** [comment]
@Particular/rabbitmq-transport-maintainers any thoughts about this one?

File: src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    /// <summary>
+    /// Extensions to <see cref="IMessageHandlerContext"/>.
+    /// </summary>
+    public static class MessageHandlerContextExtensions
+    {
+        /// <summary>
+        /// Gets the <see cref="BasicDeliverEventArgs"/> for the current message being processed.
+        /// </summary>
+        public static BasicDeliverEventArgs GetBasicDeliverEventArgs(this IMessageHandlerContext context) =>
+            context.Extensions.TryGet<BasicDeliverEventArgs>(out var args)
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs]: missing null arg check

File: src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs [hunk 2]
```diff
@@ -135,5 +135,17 @@ public static TransportExtensions<RabbitMQTransport> SetClientCertificates(this
...
+
+        /// <summary>
+        /// Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.
+        /// </summary>
+        /// <param name="transportExtensions"></param>
+        /// <param name="propagateBasicDeliverEventArgs">Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.</param>
+        /// <returns></returns>
+        public static TransportExtensions<RabbitMQTransport> PropagateBasicDeliverEventArgs(this TransportExtensions<RabbitMQTransport> transportExtensions, bool propagateBasicDeliverEventArgs)
+        {
+            transportExtensions.GetSettings().Set(SettingsKeys.PropagateBasicDeliverEventArgs, propagateBasicDeliverEventArgs);
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: missing arg null check

File: src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs [hunk 3]
```diff
@@ -135,5 +135,17 @@ public static TransportExtensions<RabbitMQTransport> SetClientCertificates(this
         /// <param name="transportExtensions"></param>
         /// <returns></returns>
         public static DelayedDeliverySettings DelayedDelivery(this TransportExtensions<RabbitMQTransport> transportExtensions) => new DelayedDeliverySettings(transportExtensions.GetSettings());
+
+        /// <summary>
+        /// Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.
+        /// </summary>
+        /// <param name="transportExtensions"></param>
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: either document the param or remove

File: src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs [hunk 4]
```diff
@@ -135,5 +135,17 @@ public static TransportExtensions<RabbitMQTransport> SetClientCertificates(this
...
+        /// <param name="propagateBasicDeliverEventArgs">Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.</param>
+        /// <returns></returns>
+        public static TransportExtensions<RabbitMQTransport> PropagateBasicDeliverEventArgs(this TransportExtensions<RabbitMQTransport> transportExtensions, bool propagateBasicDeliverEventArgs)
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: another broken fluent api.

File: src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs [hunk 5]
```diff
@@ -135,5 +135,17 @@ public static TransportExtensions<RabbitMQTransport> SetClientCertificates(this
         /// <param name="transportExtensions"></param>
         /// <returns></returns>
         public static DelayedDeliverySettings DelayedDelivery(this TransportExtensions<RabbitMQTransport> transportExtensions) => new DelayedDeliverySettings(transportExtensions.GetSettings());
+
+        /// <summary>
+        /// Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.
+        /// </summary>
+        /// <param name="transportExtensions"></param>
+        /// <param name="propagateBasicDeliverEventArgs">Specifies whether <see cref="BasicDeliverEventArgs" /> should be propagated to the <see cref="IMessageHandlerContext" /> when receiving messages.</param>
+        /// <returns></returns>
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: either document the return value or remove

File: src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿namespace NServiceBus
+{
+    using RabbitMQ.Client.Events;
+    using System;
+
+    /// <summary>
+    /// Extensions to <see cref="IMessageHandlerContext"/>.
+    /// </summary>
+    public static class MessageHandlerContextExtensions
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs]: why a new type for one method? i would just put it on `RabbitMQTransportSettingsExtensions`

File: src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        /// <summary>
+        /// Gets the <see cref="BasicDeliverEventArgs"/> for the current message being processed.
+        /// </summary>
+        public static BasicDeliverEventArgs GetBasicDeliverEventArgs(this IMessageHandlerContext context) =>
+            context.Extensions.TryGet<BasicDeliverEventArgs>(out var args)
+                ? args
+                : throw new InvalidOperationException(
+                    "BasicDeliverEventArgs have not been propagated to the message handler context." +
+                    "To enable propagation, call PropagateBasicDeliverEventArgs(true) when configuring the transport.");
```
Comment (SimonCropp) [inline src/NServiceBus.Transport.RabbitMQ/Receiving/MessageHandlerContextExtensions.cs]: `PropagateBasicDeliverEventArgs`.. i would include the type

**Comment by adamralph** [comment]
Closing, but I'll leave the branch in case we ever want to revisit this.

**Pull request closed without merging**